### ISSUE TITLE
Composite Foreign Key

### ISSUE THREAD

**Issue by gsGabriel** [dotnet/efcore#8206]
Houston, we have a problem hahahahaha

I have an issue with mapping a composite foreign key, i have two keys for the one constraint foreign key in my database.

I have one table with actions and i used this table to get a actions of anything, for the example: actions of applications or actions of the users. In this table i have a collumn type and IdEntity. In my case I can have a situation where IdEntity will repeat itself, because I'm using this column in other entities, so I created the type column, where the type will be an enum with entities. I then created a compound Foreign Key, where it corresponds to the Type and Entity Id. Currently Ef Core is not recognizing this foreing key which makes it impossible for my application to work properly. Anyone have any suggestions?

### Steps to reproduce
Generate code-first of my model

### Further technical details
EF Core version: 1.1.1
Database Provider: Npgsql.EntityFrameworkCore.PostgreSQL
IDE: Visual Studio 2017

**Comment by ajcvickers** [comment]
@gsGabriel Are you configuring the composite FK as described here: https://docs.microsoft.com/en-us/ef/core/modeling/relationships? If so, can you post a complete listing or project that reproduces what you are seeing?

**Comment by gsGabriel** [comment]
@ajcvickers Can you tell if code-First generates the relationship with foireing-key composite? My problem is precisely in this step...

My database have this constraints (i'm using postgreesql)

```
**TbActions**

Id bigint NOT NULL
IdType int NOT NULL
IdEntity bigint NOT NULL

CONSTRAINT fk_action_type_entity_users FOREIGN KEY (IdType, IdEntity)
      REFERENCES tbUsers (IdType, Id) MATCH SIMPLE
      ON UPDATE NO ACTION ON DELETE NO ACTION

CONSTRAINT fk_action_type_entity_systems FOREIGN KEY (IdType, IdEntity)
      REFERENCES tbSystems(IdType, Id) MATCH SIMPLE
      ON UPDATE NO ACTION ON DELETE NO ACTION

**TbUsers**

Id bigint NOT NULL
IdType int NOT NULL
//some data

**TbSystems**
Id bigint NOT NULL
IdType int NOT NULL
//some data
```

**Comment by ajcvickers** [comment]
@gsGabriel Please post the code you have in OnModelCreating.

**Comment by gsGabriel** [comment]
Ok @ajcvickers , but the code-first don't generate the relationship

`protected override void OnModelCreating(ModelBuilder modelBuilder)
{
    modelBuilder.Entity<TbAction>(entity =>
    {
        entity.HasKey(e => e.Id)
            .HasName("PK_tbActions");

        entity.ToTable("tbActions");

        entity.Property(e => e.IdType)
            .HasColumnName("IdType")

        entity.Property(e => e.IdEntity)
            .HasColumnName("IdEntity")
    });

    modelBuilder.Entity<TbUser>(entity =>
    {
        entity.HasKey(e => e.Id)
            .HasName("PK_tbUsers");

        entity.ToTable("tbUsers");

        entity.Property(e => e.IdType)
            .HasColumnName("IdType")
    });

    modelBuilder.Entity<TbSystem>(entity =>
    {
        entity.HasKey(e => e.Id)
            .HasName("PK_tbSystems");

        entity.ToTable("tbSystems");

        entity.Property(e => e.IdType)
            .HasColumnName("IdType")
    });
}`

**Comment by gsGabriel** [comment]
@ajcvickers 

I used the example that is in the link that you sent me, the code worked ... but the doubt stays in the code-first, should it generate the mapping or not?

**Comment by ajcvickers** [comment]
@gsGabriel Are you saying that you used the EF Core command line tools to scaffold a context and entity types from an existing database and that it didn't include the relationship? Or are you asking if conventions in EF Core should detect this relationship and use it without any configuration? If the former, then that's likely a bug in reverse engineering. If the latter, then no, EF won't detect composite keys by convention--they need to be configured explicitly.

**Comment by gsGabriel** [comment]
@ajcvickers 
I do not know I understood, but tell you what I did, I created the database and I'm trying to reverse engineer from that database. After running the command that I will put below the relationship was not implemented in OnModelCreating. The composite foreing key are not mapped.

`Scaffold-DbContext "Host=******;Port=*****;Database=db_tests;Username=postgres;Password=postgres;Persist Security Info=True" Npgsql.EntityFrameworkCore.PostgreSQL -OutputDir Context -Context ApplicationDbContext -Tables tbActions, tbUsers, tbSystems -force`

**Comment by lajones** [comment]
@gsGabriel Try adding the `-Verbose` flag to the `Scaffold-DbContext` line. That should give you more detail on what's happening.

**Comment by gsGabriel** [comment]
@lajones It just says it was compiled and it also looks up references to namespaces and assembly

**Comment by lajones** [comment]
@gsGabriel Ah, yes. Sorry - I forgot - we had a bug at that time which meant that most of the useful info from `-Verbose` wasn't coming out. :-(

Composite foreign keys are mapped in SQL Server - we have check-in tests that check that. However looking at your `OnModelCreating()` it looks like it is finding a PK on each table - but that in each case it is just `Id', not `(IdType, Id)`. This is just a guess, but there may be a problem where the FK goes looking for the key which it is supposed to be referencing on the principal end but doesn't find it and so the FK doesn't get included at all (there would be an error message about that - but as I say the useful info wasn't coming out). Can you try explicitly setting the PK of each of the 3 tables to be `(IdType, Id)` and re-running `Scaffold-DbContext`? I think that might help track it down.

Alternatively if you can try with the `2.0.0` tools then we've fixed the `-Verbose` flag problem and you would get more info on what's failing.

**Comment by gsGabriel** [comment]
Hmm ... I think I understood ... Thank you @lajones ... I'm going to take the time to contextualize the reason I'm using the composite  foreing key ... I would not need to map tbAction with tbUsers or tbSystems, my DBA allowed it not There was this relationship, but when there is no relationship the entity can not solve and put the correct Id. I'm gonna explain. My application has the Repository Pattern + UnitOfWork, that is, the SaveChanges is called at the end. So, in my code I use the Add command in the tbUsers entity and later give an Add in the tbAction entity by passing the variable id of the tbUsers entity. Subsequently I call the SaveChanges (which is encapsulated in UnitOfWork). When executing the SaveChanges the Entity Framework tries to persist the values ​​in the database however error occurs, when it tries to save the entity tbAction it does not find the Id of the tbUsers that was inserted. So the value it tries to insert is 0. I did some tests on a simpler entity in which it can "solve" the id only when there is some direct relationship. I talked to my DBA and he suggested I create the composite keys. So my question is, in order for my program to work normally and save the tbUsers Id on the tbActions entity at the time of SaveChanges do I have to have an explicit relationship?

**Comment by ajcvickers** [comment]
@lajones Do we have tests for composite FKs that reference a non-primary principal key?

**Comment by lajones** [comment]
@ajcvickers For SQL Server, yes. See e.g. [this expected result](https://github.com/aspnet/EntityFramework/blob/dev/test/EFCore.SqlServer.Design.FunctionalTests/ReverseEngineering/Expected/AllFluentApi/SqlServerReverseEngineerTestE2EContext.expected#L384). But not sure what the PostgreSQL provider does.

**Comment by ajcvickers** [comment]
@roji Looks like this could be an issue with PostgreSQL reverse engineering, although we haven't tested it yet. Any thoughts?

**Comment by ajcvickers** [comment]
@lajones Can you verify whether or not this repros with SQL Server? If it doesn't, then let's close this and open a Postgresql bug.

**Comment by gsGabriel** [comment]
@ajcvickers I discovered one possible cause of the problem. I created a test scenario where I create a new database and rebuild all tables with an sql script. When I did this the reverse engineering was correct and the composite keys were generated. I researched a bit more and realized that reverse engineering fails when there is a cache in the PostgreSQL database. For some reason it does not generate the right relationship if there are major changes in the database with larger cache. So I do not believe it's a problem in the Entity Framework...
